helix-term/src/application.rs

```diff
@@ -471,7 +473,20 @@ impl Application {
                 }
             }
             signal::SIGCONT => {
-                self.claim_term().await.unwrap();
+                // Copy/Paste from same issue from neovim:
+                // https://github.com/neovim/neovim/issues/12322
+                // https://github.com/neovim/neovim/pull/13084
+                let mut retry_count = 10;
+                while retry_count > 0 {
+                    if self.claim_term().await.is_ok() {
+                        break;
+                    }
+                    retry_count -= 1;
+                }
+                if retry_count == 0 {
+                    panic!("couldn't set terminal raw mode");
+                }
```

Comment:
consider simplifying the whole thing to the following:
```rust
let retry_count = 10;
let _ = (0..retry_count).find(|_| self.claim_term().await.is_ok()).unwrap_or_else(|| {
    panic!("Couldn't set terminal raw mode");
});
```
`find()` returns when it encounters `true` for the first time

Reply:
I don't think the functional style is clearer here. Abusing unwrap_or_else to handle the error case is really unidiomatic to me and `find` doesn't really seem intuitive for retries. The while loop is honestly quite readable here. The `rety_count == 0` check is quite intuitive. We could rename the variable `retries_left` to make it even clearer. If you desperately wanted to get rid of the while loop you could do:
``` rust
let mut retries_left = 10;
loop{
	if self.claim_term.await.is_ok(){
		break
	}
	retries_left -= 1;
	if retries_left == 0{
    	panic!("couldn't set terminal raw mode");
	}

}
```

but both versions are fine IMO

Reply:
I'm newbie in Rust and I tried to find a solution using functional style. But I'm stuck trying to use async closure in iterator.

For example this code:
```rust
let retry_count = 10;
let _ = (0..retry_count).find(|_| self.claim_term().await.is_ok()).unwrap_or_else(|| {
    panic!("Couldn't set terminal raw mode");
});```

does not compile:
```
error[E0728]: `await` is only allowed inside `async` functions and blocks
   --> helix-term/src/application.rs:481:48
    |
481 |                     .find(|_| self.claim_term().await.is_ok())
    |                           ---                  ^^^^^^ only allowed inside `async` functions and blocks
    |                           |
    |                           this is not `async`
    ```

Reply:
You'd need to have an async closure there with 
```rs
async |_| self.claim_term().await.is_ok()
```
but as those are currently unstable, it would need some extra trickery to get working, which would be way more trouble than it's worth.

Even if the looping approach is a bit messy due to having the extra mutable variable, it's probably the simplest way to do this while having the panic functionality in place. 

Pascal's approach could be combined with what I originally suggested to be more concise
```rs
for retries in 1..=10 {
    if self.claim_term.await.is_ok(){
        break
    }
    if retries == 10 {
        panic!("couldn't set terminal raw mode");
    }
}
```
But with that you're going to be doing that second if check every iteration, even though it's redundant for 9 of those.